src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -2181,32 +2420,37 @@ public static TSource Max<TSource>(this IEnumerable<TSource> source)
             TSource value = default(TSource);
             if (value == null)
             {
-                foreach (TSource x in source)
+                using (IEnumerator<TSource> e = source.GetEnumerator())
                 {
-                    if (x != null && (value == null || comparer.Compare(x, value) > 0))
-                        value = x;
+                    if (!e.MoveNext()) return value;
+                    value = e.Current;
+                    // A do…while may seem more natural here, but profiled worse across a wide set of test data
```

Comment:
I tried a great many variants, but I did consider null-free and low-null to be more important. I also thought consistency between the methods more important than if a micro-opt was available by taking one approach for one datatype and another approach for another.